XiEditor/StatusBar.swift

```diff
@@ -0,0 +1,140 @@
+        self.leadingAnchor.constraint(equalTo: editView.leadingAnchor).isActive = true
+        self.trailingAnchor.constraint(equalTo: editView.trailingAnchor).isActive = true
+        self.bottomAnchor.constraint(equalTo: editView.bottomAnchor).isActive = true
+    }
+
+    // Adds a status bar item. Only appends status bar items for now.
+    func addStatusItem(_ item: StatusItem) {
+        // If item exists, update its value
+        if let existingItem = currentItems[item.key] {
+            updateStatusItem(existingItem.key, item.value)
+            return
+        }
+        print("added item \(item.key)")
+        item.translatesAutoresizingMaskIntoConstraints = false
+        item.textColor = NSColor.black
+        self.addSubview(item)
+        item.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
+        currentItems[item.key] = item
+        
+        switch item.barAlignment {
+            case .left:
+                leftItems.append(item)
+            case .right:
+                rightItems.append(item)
+        }
+        self.needsUpdateConstraints = true
+    }
+
+    // Update a status bar item with a new value.
+    func updateStatusItem(_ key: String, _ value: String) {
+        if let item = currentItems[key] {
+            item.stringValue = value
+        }
+    }
+
+    // Removes status bar item with a specified key.
+    func removeStatusItem(_ key: String) {
+        if let item = currentItems[key]  {
+            item.removeFromSuperview()
+            leftItems = leftItems.filter { $0 != item }
+            rightItems = rightItems.filter { $0 != item }
+            currentItems.removeValue(forKey: key)
+        }
+        self.needsUpdateConstraints = true
+    }
+
+    // Update constraints of status bar items.
+    // Called when the status bar item state is modified.
+    override func updateConstraints() {
+        for item in currentItems.values {
+            if item.barAlignment == .left {
+                if item == leftItems.first {
+                    item.leadingAnchor.constraint(equalTo: self.leadingAnchor).isActive = true
+                } else {
+                    item.leadingAnchor.constraint(equalTo:
+                        leftItems[leftItems.index(of: item)! - 1].trailingAnchor, constant: 10).isActive = true
+                }
+            } else {
+                if item == rightItems.first {
+                    item.trailingAnchor.constraint(equalTo: self.trailingAnchor).isActive = true
+                } else {
+                    item.trailingAnchor.constraint(equalTo:
+                        rightItems[rightItems.index(of: item)! - 1].leadingAnchor, constant: -10).isActive = true
+                }
+            }
+        }
+        super.updateConstraints()
+    }
+
+    override func draw(_ dirtyRect: NSRect) {
+        super.draw(dirtyRect)
+        let path = NSBezierPath()
+        backgroundColor.setFill()
+        __NSRectFill(dirtyRect)
```

Comment:
I... don't think we're supposed to use this function. I think there's just a `fill()` method on NSRect now?